analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs

```diff
@@ -141,59 +141,35 @@ ReportDiagnostic Severity(string id)
             }
         }
 
-        private static void CompareActualToExpected(string languageVersion, Diagnostic[] diagnostics, FileIssueLocations[] expectedIssuesPerFile, bool compareIdToMessage)
+        private static void CompareActualToExpected(Diagnostic[] diagnostics, CompilationIssues expected, bool compareIdToMessage)
         {
-            var actualIssues = new CompilationIssues(languageVersion, diagnostics);
-            actualIssues.Dump();
-
-            // ToDo: Throw this away
-            foreach (var diagnostic in diagnostics.OrderBy(x => x.Location.SourceSpan.Start))
+            var actual = new CompilationIssues(expected.LanguageVersion, diagnostics);
+            var pairs = MatchPairs(actual, expected);
+            var assertionMessages = new StringBuilder();
+            string previousFilePath = null;
+            foreach (var pair in pairs.OrderBy(x => x.FilePath).ThenBy(x => x.LineNumber).ThenBy(x => x.Start))
             {
-                var expectedIssues = ExpectedIssues(expectedIssuesPerFile, diagnostic.Location);
-                var issueId = VerifyPrimaryIssue(languageVersion,
-                    expectedIssues,
-                    issue => issue.IsPrimary,
-                    diagnostic.Location,
-                    compareIdToMessage ? diagnostic.Id : diagnostic.GetMessage(),
-                    compareIdToMessage
-                        ? $"{languageVersion}: Unexpected build error [{diagnostic.Id}]: {diagnostic.GetMessage()} on line {diagnostic.Location.GetLineNumberToReport()}"
-                        : null);
-
-                var secondaryLocations = diagnostic.AdditionalLocations
-                    .Select((_, i) => diagnostic.GetSecondaryLocation(i))
-                    .OrderBy(x => x.Location.GetLineNumberToReport())
-                    .ThenBy(x => x.Location.GetLineSpan().StartLinePosition.Character);
-
-                foreach (var secondaryLocation in secondaryLocations)
+                if (previousFilePath != pair.FilePath)
                 {
-                    var expectedIssuesSecondaryLocation = ExpectedIssues(expectedIssuesPerFile, secondaryLocation.Location);
-                    VerifySecondaryIssue(languageVersion,
-                        expectedIssuesSecondaryLocation,
-                        issue => issue.IssueId == issueId && !issue.IsPrimary,
-                        secondaryLocation.Location,
-                        secondaryLocation.Message,
-                        issueId);
+                    if (assertionMessages.Length > 0)
+                    {
+                        assertionMessages.AppendLine();
```

Comment:
I would move this at the end of the loop, and remove the Length-check completely for simplicity.
Worst-case you add an extra new-line at the end.